File: Java/WordSearch.java
public class WordSearch {
	// With additional space
	private boolean dfs(char[][] board, int r, int c, String word, int i, boolean[][] used) {
        if ((r < 0) || (r >= board.length) || (c < 0) || (c >= board[r].length) 
            || used[r][c]) {
            return false;
        }
        if (board[r][c] != word.charAt(i)) {
            return false;
        }
        if (i == word.length() - 1) {
            return true;
        }
        used[r][c] = true;
        boolean exist = dfs(board, r-1, c, word, i+1, used) || dfs(board, r, c+1, word, i+1, used)
         || dfs(board, r+1, c, word, i+1, used) || dfs(board, r, c-1, word, i+1, used);
        used[r][c] = false;
        return exist;
    }
    
    public boolean exist(char[][] board, String word) {
        boolean[][] used = new boolean[board.length][board[0].length];
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (dfs(board, i, j, word, 0, used)) {
                    return true;
                }
            }
        }
        return false;
    }

    // Without additional space
    private boolean dfs2(char[][] board, int r, int c, String word, int i) {
        if ((r < 0) || (r >= board.length) || (c < 0) || (c >= board[r].length) 
            || (board[r][c] == '#')) {
            return false;
        }
        if (board[r][c] != word.charAt(i)) {
            return false;
        }
        if (i == word.length() - 1) {
            return true;
        }
        char character = board[r][c];
        board[r][c] = '#'; // another simpler way: board[r][c] ^= 256;
        boolean exist = dfs(board, r-1, c, word, i+1) || dfs(board, r, c+1, word, i+1)
         || dfs(board, r+1, c, word, i+1) || dfs(board, r, c-1, word, i+1);
        board[r][c] = character; // board[r][c] ^= 256
        return exist;
    }
    
    // Better implementation of above dfs function:
    private boolean dfs(char[][] board, int i, int j, String word, int wordId) {
        if (wordId >= word.length()) {
            return true;
        }
        if (i < 0 || j < 0 || i >= board.length || j >= board[i].length || board[i][j] != word.charAt(wordId)) {
            return false;
        }
        char c = board[i][j];
        board[i][j] = '#';
        if (dfs(board, i - 1, j, word, wordId + 1) || dfs(board, i + 1, j, word, wordId + 1)
         || dfs(board, i, j - 1, word, wordId + 1) || dfs(board, i, j + 1, word, wordId + 1)) {
            return true;
        }
        board[i][j] = c;
        return false;
    }

    public boolean exist2(char[][] board, String word) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (dfs(board, i, j, word, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: C++/WordSearch.cc
#include <deque>
#include <iostream>
#include <vector>

using namespace std;

// Recursive solution using DFS. Pay attention to this kind of
// multi-starting-points problem. Possible improvement: instead of using the
// visited array, we can use '#' to mark the visited points on the board and
// change it back after the recursive search.
//
// Note that iterative verstion of DFS is non-trivial, since in each state
// there must be a structure storing the visited information, which in this
// case, is same size of the board.
bool dfsWord0(vector<vector<char> >&board, int len, pair<int, int>& pos, string& word, vector<vector<bool> >& visited)
{
    int w = pos.first;
    int l = pos.second;
    int wmax = board.size();
    int lmax = board[0].size();

    if (board[w][l] != word[len])
        return false;
    else 
        visited[w][l] = true;

    if (len == word.size() - 1) {
        return true;
    }

    pair<int, int> adjs[4] = {{w+1, l}, {w-1, l}, {w, l+1}, {w, l-1}};
    for (int i = 0; i < 4; ++i)
    {
        pair<int, int> cur = adjs[i];
        if ((cur.first >= 0) && (cur.first < wmax) && (cur.second >= 0) && (cur.second < lmax)
                 && !visited[cur.first][cur.second]) {
           if (dfsWord0(board, len+1, cur, word, visited))
               return true;
        }
    }
    visited[w][l] = false;
    return false;
}

bool exist0(vector<vector<char> > &board, string word) {
    vector<vector<bool> > visited(board.size(), vector<bool>(board[0].size(), false));
    
    if (word.size() == 0)
        return false;

    for (int i = 0; i < board.size(); ++i)
    {
        for (int j = 0; j < board[0].size(); ++j)
        {
            pair<int, int> curPos(i, j);
            if (dfsWord(board, 0, curPos, word, visited))
                return true;
        }
    }

    return false;
}

//Without additional array
bool dfsExist(vector<vector<char> >& board, string& word, int idWord, int rBoard, int cBoard)
{
    int nr = board.size();
    int nc = board[0].size();
    pair<int, int> adjs[4] = {{rBoard, cBoard+1}, {rBoard, cBoard-1}, {rBoard+1, cBoard}, {rBoard-1, cBoard}};
    
  
        
    if (word[idWord] != board[rBoard][cBoard])
        return false;
 
    if (idWord == word.size() - 1) 
        return true;
    char c = board[rBoard][cBoard];
    board[rBoard][cBoard] = '#';
    
    for (int i = 0; i < 4; ++i)
    {
        int r = adjs[i].first;
        int c = adjs[i].second;
        if ((r >= 0) && (r < nr) && (c >= 0) && (c < nc)) {
         
            if (dfsExist(board, word, idWord+1, r, c))
                return true;
        
        }
    }
    
    board[rBoard][cBoard] = c;
    
    return false;
}

bool exist(vector<vector<char> > &board, string word) {
    int nr = board.size();
    if (nr == 0)
        return false;
    int nc = board[0].size();
    if (nc == 0)
        return false;
        
    for (int i = 0; i < nr; ++i)
    {
        for (int j = 0; j < nc; ++j)
        {
            if (dfsExist(board, word, 0, i, j))
                return true;
        }
    }
    
    return false;
}

// BFS is not good for this problem. 
// 1. Board = AA
//            AA,  word = "AAAA", BFS won't work.
// And to make it work, we need to modify the judging repetition part.
// 2. Not easy to store the visited points. Cannot use board to store them.


int main(int argc, char** argv)
{
    vector<vector<char> > board = {
        {'A', 'B', 'C', 'E'},
        {'S', 'F', 'C', 'S'},
        {'A', 'D', 'E', 'E'}
    };
    
    vector<string> tests = {"ABCCED", "SEE", "ABCB"};
/*    vector<vector<char> > board = {
        {'A', 'A'}
    };
    vector<string> tests = {"AAA"};*/
    for (int i = 0; i < tests.size(); ++i)
    {
        cout << exist(board, tests[i]) << endl; 
    }
    return 0;
}
